### PR TITLE
feat: sets up offset logic on `nippy-jar`

### PR DESCRIPTION
**Pull request by joshieDo**
Adds the real offset logic and brings everything together.

There are two offset lists:
* `offsets`: holds a sorted list of the data file offsets. 
  * example: `[row0_col0_value_offset, row0_col1_value_offset, row1_col0_value_offset,... rowN_colM_value_offset...]`
  * useful for `transaction_by_number`
* `offsets_index`: this is an index `PHF(value)-(random integer)->row_number`, which representation is an unsorted integer list.
  * example: `func(row0)` might return `2`, while `func(row1)` might return `0`. So the output will be: `[ row1, X, row0, ..., rowN]`  Value doesn't have to be part of the data. (imagine having a file with transactions, but not actually storing their hashes)
  * useful for `transaction_by_hash`

TL;DR can be seen implemented on `test_full_nippy_jar`, `NippyJarCursor::row_by_filter`

Also, refactored a few things and added some additional docs.

#4601 should be merged first (this is one is pointing to that branch)

### CODE REVIEW

File: crates/storage/nippy-jar/src/cursor.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+use std::{clone::Clone, fs::File, sync::Mutex};
+use sucds::int_vectors::Access;
+use zstd::bulk::Decompressor;
+
+/// Simple cursor implementation to retrieve data from [`NippyJar`].
+pub struct NippyJarCursor<'a> {
+    /// [`NippyJar`] which holds most of the required configuration to read from the file.
+    jar: &'a NippyJar,
+    /// Optional dictionary decompressors.
+    zstd_decompressors: Option<Mutex<Vec<Decompressor<'a>>>>,
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:12]: why do we need a mutex here?

File: crates/storage/nippy-jar/src/cursor.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                let next_value_offset =
+                    self.jar.offsets.select(offset_pos + 1).expect("should exist");
+                &self.mmap_handle[value_offset..next_value_offset]
+            };
+
+            // Decompression
+            if let Some(zstd_dict_decompressors) = &self.zstd_decompressors {
+                // Uses zstd dictionaries
+                let extra_capacity =
+                    (column_value.len() * 2).saturating_sub(self.tmp_buf.capacity());
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:116]: can we add a comment why we reserve 2 times more than required here?

File: crates/storage/nippy-jar/src/cursor.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    tmp_buf: Vec<u8>,
+    /// Cursor row position.
+    row: u64,
+    /// Cursor column position.
+    col: u64,
+}
+
+impl<'a> std::fmt::Debug for NippyJarCursor<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NippyJarCursor {{ config: {:?} }}", self.jar)
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:29]:
```suggestion
        f.debug_struct("NippyJarCursor")
            .field("config", &self.jar)
            .finish_non_exhaustive()
```

File: crates/storage/nippy-jar/src/cursor.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        self.col = 0;
+    }
+
+    /// Returns a row, searching it by an entry used during [`NippyJar::prepare_index`].
+    ///
+    /// **May return false positives.**
+    ///
+    /// Example usage would be querying a transactions file with a transaction hash which is **NOT**
+    /// stored in file.
+    pub fn row_by_filter(&mut self, value: &[u8]) -> Result<Option<Vec<Vec<u8>>>, NippyJarError> {
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:63]: nit: return type might be a bit hard to grasp because of nested vecs, may be worth introducing an alias like `type Value = Vec<u8>` and mentioning in the comments that it returns columns

File: crates/storage/nippy-jar/src/cursor.rs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                if let Some(row_index) = phf.get_index(value)? {
+                    self.row =
+                        self.jar.offsets_index.access(row_index as usize).expect("built from same")
+                            as u64;
+                    return self.next_row()
+                }
+            }
+        }
+
+        Ok(None)
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:79]: should we return an error here in case when passed nippyjar doesn't have a filter and phf prepared for this type of query?

File: crates/storage/nippy-jar/src/cursor.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    /// Returns a row by its number.
+    pub fn row_by_number(&mut self, row: usize) -> Result<Option<Vec<Vec<u8>>>, NippyJarError> {
+        self.row = row as u64;
+        self.next_row()
+    }
+
+    /// Advances cursor to next row and returns it.
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:88]: Doesn't it return the current row, and only then advances the cursor?

File: crates/storage/nippy-jar/src/cursor.rs (hunk 4 above)
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:63]: nit: method and argument names are a bit confusing, maybe something like `row_by_key` and `key: &[u8]`? IIUC with `transaction_by_hash` as an example: it accepts the hash, checks the filter, resolves into a offset index and gets the offset. Lookup by filter seems like an internal implementation detail for me here.

File: crates/storage/nippy-jar/src/cursor.rs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    file_handle: File,
+    /// Data file.
+    mmap_handle: Mmap,
+    /// Temporary buffer to unload data to (if necessary), without reallocating memory on each
+    /// retrieval.
+    tmp_buf: Vec<u8>,
+    /// Cursor row position.
+    row: u64,
+    /// Cursor column position.
+    col: u64,
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/cursor.rs:24]: do we use the column anywhere?

File: crates/storage/nippy-jar/src/cursor.rs (hunk 4 above)
Comment (joshieDo) [reply crates/storage/nippy-jar/src/cursor.rs:63]: `key` feels like a misnomer though, but i agree overall

File: crates/storage/nippy-jar/src/cursor.rs (hunk 7 above)
Comment (joshieDo) [reply crates/storage/nippy-jar/src/cursor.rs:24]: has already been removed in #4642

File: crates/storage/nippy-jar/src/cursor.rs (hunk 4 above)
Comment (shekhirin) [reply crates/storage/nippy-jar/src/cursor.rs:63]: hmm, but isn't the interface basically resembles a `BTreeMap` where key is the transaction hash and you can retrieve the value with `row_by_filter`, and you can also get the value by its index with `row_by_number`?

File: crates/storage/nippy-jar/src/cursor.rs (hunk 1 above)
Comment (joshieDo) [reply crates/storage/nippy-jar/src/cursor.rs:12]: decompressor needs to be mutable to decompress... which is less than ideal for sharing it across places. this is temporary, and will be changed in the future. (there's a missing Arc)

Comment (shekhirin) [reply crates/storage/nippy-jar/src/cursor.rs:12]: makes sense, but we already take `&mut self` when using decompressor, so we should be good?

File: crates/storage/nippy-jar/src/cursor.rs [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    }
+}
+
+impl<'a> NippyJarCursor<'a> {
+    pub fn new(
+        config: &'a NippyJar,
+        zstd_decompressors: Option<Vec<Decompressor<'a>>>,
+    ) -> Result<Self, NippyJarError> {
+        let file = File::open(config.data_path())?;
+        let mmap = unsafe { Mmap::map(&file)? };
```
Comment (mattsse) [inline crates/storage/nippy-jar/src/cursor.rs:40]: SAFETY?

File: crates/storage/nippy-jar/src/cursor.rs [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        row: usize,
+    ) -> Result<Option<Row>, NippyJarError> {
+        self.row = row as u64;
+        self.next_row_with_cols::<MASK, COLUMNS>()
+    }
+
+    /// Returns the current value and advances the row.
+    ///
+    /// Uses a `MASK` to only read certain columns from the row.
+    pub fn next_row_with_cols<const MASK: usize, const COLUMNS: usize>(
```
Comment (mattsse) [inline crates/storage/nippy-jar/src/cursor.rs:147]: how many different combinations are we expecting? this could generate a lot of code 

and I assume the benefit of having this const is neglible?

File: crates/storage/nippy-jar/src/cursor.rs (hunk 8 above)
Comment (joshieDo) [reply crates/storage/nippy-jar/src/cursor.rs:42]: added

**Review by mattsse — APPROVED** [review]
